### DISCUSSION TITLE
airflow remote logging with gcs not working

### DISCUSSION THREAD

**Discussion by alagappan-al** [apache/airflow#27807] [Q&A]
I am trying to setup the airflow using the docker-compose
https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html

The environment is up and running.  Modified airflow.cfg to enable remote logging
[logging]
base_log_folder = /opt/airflow/logs
remote_logging = True
remote_log_conn_id = google_cloud_default
google_key_path = /home/airflow/keyfile.json
remote_base_log_folder = gs://xxxxxxxx-bucket/logs

Verified google_cloud_default connection with other dags. It has enough permission to read/write from the bucket. Recreated containers after modifying the airflow.cfg. 

Error message: 
Unable to read remote log from gs://
404 GET [https://storage.googleapis.com/download/storage/v1/b/](https://storage.googleapis.com/download/storage/v1/b/)

I followed official documentation, stack overflow suggestions and git hub discussions. Still not able to achieve the results. Do I need to create custom folder with dag_id/run_id in gcp bucket? What else do we have to add in the airflow.cfg? any help is appreciated.

Airflow version: 2.3.3

**Comment by alagappan-al** [comment]
@potiuk @Taragolis

**Reply by potiuk** [reply]
I think your problem is that the service account you have: /home/airflow/keyfile.json has not enough access. 

https://airflow.apache.org/docs/apache-airflow-providers-google/stable/logging/gcs.html

> By default Application Default Credentials are used to obtain credentials. You can also set google_key_path option in [logging] section, if you want to use your own service account.

The ApplicationDefaultCredentials - is the standard way how google libraries retrieve credentials https://cloud.google.com/docs/authentication/application-default-credentials. This is different than `google_cloud_default` - this keyfile.json is NOT 'google_cloud_default` - it is only used to read/write the logs. 

If you override the "default" credentals by providing the .json file, you need to make sure you can read/write the files using that key.

The main reason for this is (I guess) security - that this should be very limited account, only capable of reading/writing the logs, where google_default_credentials are usually much more powerful as they are used to actually run some services in the cloud.

**Reply by potiuk** [reply]
BTW. I clarified it a bit better in our docs: https://github.com/apache/airflow/pull/28084/ to make it clear that it has nothing to do with `google_cloud_default`

**Reply by alagappan-al** [reply]
Hi @potiuk 
Thanks for taking look into this. The key file I am using has enough permission. I created the google_cloud_default connection using the credentials. I am using the same connection to read/write back to GCS bucket in Google Managed airflow (Cloud Composer) Dags. I am having problems when I use the same in the airflow setup using docker composer.

**Reply by potiuk** [reply]
Did you try access to logs from inside the docker? 

Unfortunately it is likely something that you have to look at whether everything works from your workers inside your compose. Cannot help you more - you need to debug it in more detail. If your "gcloud command" using the key you configured (and hopefully made available to your container) works from the container, your log integration should work.

Just make absolutely sure that you run your tests in exactly the same circumstances as your airflow - same container (with exec-ing into actual container), same user, same environment, same everything. It's something that you will need to do anyway in the future if you want to be able to debug and solve other problems in the future.

By choosing your docker-compose setup as copy of your compose, you take on the burden on making sure that you can esily reproduce it.

BTW. Note that in Google Composer, you "Application Default Credentials" are using something completely different by default. They are authenticated using Metadata server on your instances and service accounts that are present on Composer machines. So I think you simply need to take a closer look and make sure that the authentication you use is actually the right one.

You can also look at https://cloud.google.com/docs/authentication/application-default-credentials if you have problem with mounting and making sure that your `google_key_path` works as intended. There are several ways application-default-credentials can read authentication (including environment variables) and you can try all of them.
